include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,803 @@
+        }
+
+    private:
+        alignas(ICommand) uint8_t m_data[STORAGE_SIZE];
+
+        void wipeNextCommandSize()
+        {
+            const auto cursor = header.m_commandAllocator.get_allocated_size();
+            // This also wipes the vtable ptr.
+            const uint32_t wipeSize = offsetof(IGPUCommandPool::ICommand, m_size) + sizeof(IGPUCommandPool::ICommand::m_size);
+            if (cursor + wipeSize < header.m_commandAllocator.get_total_size())
+                memset(m_data + cursor, 0, wipeSize);
+        }
+    };
+    static_assert(sizeof(CCommandSegment) == COMMAND_SEGMENT_SIZE);
+
+    class CBeginCmd;
+    class CBindIndexBufferCmd;
+    class CDrawIndirectCmd;
+    class CDrawIndexedIndirectCmd;
+    class CDrawIndirectCountCmd;
+    class CDrawIndexedIndirectCountCmd;
+    class CBeginRenderPassCmd;
+    class CPipelineBarrierCmd;
+    class CBindDescriptorSetsCmd;
+    class CBindComputePipelineCmd;
+    class CUpdateBufferCmd;
+    class CResetQueryPoolCmd;
+    class CWriteTimestampCmd;
+    class CBeginQueryCmd;
+    class CEndQueryCmd;
+    class CCopyQueryPoolResultsCmd;
+    class CBindGraphicsPipelineCmd;
+    class CPushConstantsCmd;
+    class CBindVertexBuffersCmd;
+    class CCopyBufferCmd;
+    class CCopyBufferToImageCmd;
+    class CBlitImageCmd;
+    class CCopyImageToBufferCmd;
+    class CExecuteCommandsCmd;
+    class CDispatchIndirectCmd;
+    class CWaitEventsCmd;
+    class CCopyImageCmd;
+    class CResolveImageCmd;
+    class CClearColorImageCmd;
+    class CClearDepthStencilImageCmd;
+    class CFillBufferCmd;
+    class CSetEventCmd;
+    class CResetEventCmd;
+    class CWriteAccelerationStructurePropertiesCmd;
+    class CBuildAccelerationStructuresCmd; // for both vkCmdBuildAccelerationStructuresKHR and vkCmdBuildAccelerationStructuresIndirectKHR
+    class CCopyAccelerationStructureCmd;
+    class CCopyAccelerationStructureToOrFromMemoryCmd; // for both vkCmdCopyAccelerationStructureToMemoryKHR and vkCmdCopyMemoryToAccelerationStructureKHR
+
+    IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx)
+        : IBackendObject(std::move(dev)), m_commandSegmentPool(COMMAND_SEGMENTS_PER_BLOCK* COMMAND_SEGMENT_SIZE, 0u, MAX_COMMAND_SEGMENT_BLOCK_COUNT, MIN_POOL_ALLOC_SIZE),
+        m_flags(_flags), m_familyIx(_familyIx)
+    {}
+
+    inline core::bitflag<E_CREATE_FLAGS> getCreationFlags() const { return m_flags; }
+    inline uint32_t getQueueFamilyIndex() const { return m_familyIx; }
+
+    // OpenGL: nullptr, because commandpool doesn't exist in GL (we might expose the internal allocator in the future)
+    // Vulkan: const VkCommandPool*
+    virtual const void* getNativeHandle() const = 0;
+
+    template <typename Cmd, typename... Args>
+    Cmd* emplace(CCommandSegment::Iterator& segmentListHeadItr, CCommandSegment*& segmentListTail, Args&&... args)
```

Comment:
> Fix the inversion of control happening between IGPUCommandPool and IGPUCommandBuffer. https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#issuecomment-1193632400

So in the end you didn't make them?

Reply:
I didn't because you would need different wrapper methods for different segmented tapes (since they have different head and tail pointers).